File: frontend/services/patientService.ts
import { API_ENDPOINTS } from "./api";
import { getAuthHeaders } from "@/lib/auth";
import log from "@/lib/logger";
import type {
  Patient,
  CreatePatientRequest,
  CreatePatientResponse,
  PatientListResponse,
  TimelineStage,
  TimelineWithDetail,
  CreateTimelineRequest,
  CreateTimelineResponse,
  CreateTimelineDetailRequest,
  CreateMedicalImageRequest,
  BatchCreateMetricsRequest,
  PatientTodo,
  CreateTodoRequest,
  UpdateTodoStatusRequest,
  ApiSuccessResponse,
  TimelineListResponse,
  TimelineDetailResponse,
  TodoListResponse,
  LabReport,
  ImagingReport,
  CreateLabReportRequest,
  CreateImagingReportRequest,
} from "@/types/patient";
 
 
// ============================================================================
// Patient Basic Info Services
// ============================================================================
/**
 * Create a new patient record
 */
export async function createPatient(
  patientData: CreatePatientRequest
): Promise<CreatePatientResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.create, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(patientData),
    });


    if (!response.ok) {
      throw new Error(`Failed to create patient: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to create patient:", error);
    throw error;
  }
}
 
/**
 * Get patient by ID
 */
export async function getPatient(patientId: number): Promise<Patient> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.detail(patientId), {
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get patient: ${response.statusText}`);
    }

    const data = await response.json();
    return data.patient;
  } catch (error) {
    log.error(`Failed to get patient ${patientId}:`, error);
    throw error;
  }
}

/**
 * Get patient by email address (for patient portal)
 */
export async function getPatientByEmail(email: string): Promise<Patient> {
  try {
    const response = await fetch(`${API_ENDPOINTS.patient.list.replace('/list', '/profile/by_email')}?email=${encodeURIComponent(email)}`, {
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get patient by email: ${response.statusText}`);
    }

    const data = await response.json();
    return data.patient;
  } catch (error) {
    log.error(`Failed to get patient by email ${email}:`, error);
    throw error;
  }
}
 
/**
 * List patients with optional search and filters
 */
export async function listPatients(params?: {
  search?: string;
  filter_type?: string;
  limit?: number;
  offset?: number;
}): Promise<PatientListResponse> {
  try {
    const queryParams = new URLSearchParams();
    if (params?.search) queryParams.append("search", params.search);
    if (params?.filter_type) queryParams.append("filter_type", params.filter_type);
    if (params?.limit) queryParams.append("limit", params.limit.toString());
    if (params?.offset) queryParams.append("offset", params.offset.toString());
 
    const url = `${API_ENDPOINTS.patient.list}${queryParams.toString() ? `?${queryParams.toString()}` : ""}`;

    const response = await fetch(url, {
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      let errorMessage = `Failed to list patients: ${response.status} ${response.statusText}`;
      try {
        const errorData = await response.json();
        if (errorData.detail) {
          errorMessage = `Failed to list patients (${response.status}): ${errorData.detail}`;
        } else if (errorData.message) {
          errorMessage = `Failed to list patients (${response.status}): ${errorData.message}`;
        } else {
          errorMessage = `Failed to list patients (${response.status}): ${JSON.stringify(errorData)}`;
        }
      } catch (e) {
        // If response is not JSON, use status text
        const text = await response.text().catch(() => "");
        if (text) {
          errorMessage = `Failed to list patients (${response.status}): ${text}`;
        }
      }
      throw new Error(errorMessage);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to list patients:", error);
    // Ensure error message is a string
    if (error instanceof Error) {
      // Check for network errors
      if (error.message.includes("Failed to fetch") || error.message.includes("NetworkError")) {
        throw new Error("无法连接到服务器，请检查后端服务是否正在运行");
      }
      throw error;
    } else if (typeof error === "string") {
      throw new Error(error);
    } else {
      throw new Error(`Failed to list patients: ${JSON.stringify(error)}`);
    }
  }
}
 
/**
 * Update patient information
 */
export async function updatePatient(
  patientId: number,
  patientData: Partial<CreatePatientRequest>
): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.update(patientId), {
      method: "PUT",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(patientData),
    });
 
    if (!response.ok) {
      let errorMessage = `Failed to update patient: ${response.statusText}`;
      try {
        const errorData = await response.json();
        if (errorData.detail) {
          errorMessage = `Failed to update patient (${response.status}): ${errorData.detail}`;
        } else if (errorData.message) {
          errorMessage = `Failed to update patient (${response.status}): ${errorData.message}`;
        }
      } catch (e) {
        // If response is not JSON, use status text
        const text = await response.text().catch(() => "");
        if (text) {
          errorMessage = `Failed to update patient (${response.status}): ${text}`;
        }
      }
      throw new Error(errorMessage);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to update patient ${patientId}:`, error);
    throw error;
  }
}
 
/**
 * Delete patient (soft delete)
 */
export async function deletePatient(patientId: number): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.delete(patientId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to delete patient: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete patient ${patientId}:`, error);
    throw error;
  }
}
 
// ============================================================================
// Timeline Services
// ============================================================================
 
/**
 * Create a new timeline stage
 */
export async function createTimelineStage(
  timelineData: CreateTimelineRequest
): Promise<CreateTimelineResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.timeline.create, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(timelineData),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to create timeline: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to create timeline:", error);
    throw error;
  }
}
 
/**
 * Get patient timeline
 */
export async function getPatientTimeline(patientId: number): Promise<TimelineStage[]> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.timeline.list(patientId), {
      headers: getAuthHeaders(),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to get timeline: ${response.statusText}`);
    }
 
    const data: TimelineListResponse = await response.json();
    return data.timelines;
  } catch (error) {
    log.error(`Failed to get timeline for patient ${patientId}:`, error);
    throw error;
  }
}
 
/**
 * Get timeline detail for a specific stage
 */
export async function getTimelineDetail(timelineId: number): Promise<TimelineWithDetail> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.timeline.detail(timelineId), {
      headers: getAuthHeaders(),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to get timeline detail: ${response.statusText}`);
    }
 
    const data: TimelineDetailResponse = await response.json();
    return data.timeline;
  } catch (error) {
    log.error(`Failed to get timeline detail ${timelineId}:`, error);
    throw error;
  }
}
 
/**
 * Save timeline detail (create or update)
 */
export async function saveTimelineDetail(
  detailData: CreateTimelineDetailRequest
): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.timeline.createDetail, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(detailData),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to save timeline detail: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to save timeline detail:", error);
    throw error;
  }
}

 

/**
 * Delete a timeline stage
 */
export async function deleteTimeline(timelineId: number): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.timeline.delete(timelineId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to delete timeline: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete timeline ${timelineId}:`, error);
    throw error;
  }
}

 
/**
 * Delete all images for a timeline
 */
export async function deleteTimelineImages(timelineId: number): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.image.delete(timelineId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to delete timeline images: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete timeline images ${timelineId}:`, error);
    throw error;
  }
}
 
/**
 * Delete all metrics for a timeline
 */
export async function deleteTimelineMetrics(timelineId: number): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.metrics.delete(timelineId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to delete timeline metrics: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete timeline metrics ${timelineId}:`, error);
    throw error;
  }
}

// ============================================================================
// Medical Image Services
// ============================================================================
/**
 * Create a medical image record
 */
export async function createMedicalImage(
  imageData: CreateMedicalImageRequest
): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.image.create, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(imageData),
    });

    if (!response.ok) {
      throw new Error(`Failed to create medical image: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to create medical image:", error);
    throw error;
  }
}


// ============================================================================
// Metrics Services
// ============================================================================

/**
 * Batch create patient metrics
 */
export async function batchCreateMetrics(
  metricsData: BatchCreateMetricsRequest
): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.metrics.batchCreate, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(metricsData),
    });

    if (!response.ok) {
      throw new Error(`Failed to create metrics: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to batch create metrics:", error);
    throw error;
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

const getJsonAuthHeaders = () => ({
  ...getAuthHeaders(),
  "Content-Type": "application/json",
});

async function jsonRequest<T>(url: string, init: RequestInit, errorPrefix: string): Promise<T> {
  const response = await fetch(url, init);

  if (!response.ok) {
    throw new Error(`${errorPrefix}: ${response.statusText}`);
  }

  return response.json();
}

// ============================================================================
// Todo Services
// ============================================================================

export async function createPatientTodo(
  todoData: CreateTodoRequest
): Promise<ApiSuccessResponse> {
  try {
    return await jsonRequest<ApiSuccessResponse>(
      API_ENDPOINTS.patient.todo.create,
      {
        method: "POST",
        headers: getJsonAuthHeaders(),
        body: JSON.stringify(todoData),
      },
      "Failed to create todo"
    );
  } catch (error) {
    log.error("Failed to create todo:", error);
    throw error;
  }
}

export async function getPatientTodos(
  patientId: number,
  status?: string
): Promise<PatientTodo[]> {
  try {
    const statusQuery = status ? `?status=${encodeURIComponent(status)}` : "";
    const data = await jsonRequest<TodoListResponse>(
      API_ENDPOINTS.patient.todo.list(patientId) + statusQuery,
      {
        headers: getAuthHeaders(),
      },
      "Failed to get todos"
    );

    return data.todos;
  } catch (error) {
    log.error(`Failed to get todos for patient ${patientId}:`, error);
    throw error;
  }
}

export async function updateTodoStatus(
  todoId: number,
  statusData: UpdateTodoStatusRequest
): Promise<ApiSuccessResponse> {
  try {
    return await jsonRequest<ApiSuccessResponse>(
      API_ENDPOINTS.patient.todo.updateStatus(todoId),
      {
        method: "PUT",
        headers: getJsonAuthHeaders(),
        body: JSON.stringify(statusData),
      },
      "Failed to update todo status"
    );
  } catch (error) {
    log.error(`Failed to update todo status ${todoId}:`, error);
    throw error;
  }
}

export async function deletePatientTodo(todoId: number): Promise<ApiSuccessResponse> {
  try {
    return await jsonRequest<ApiSuccessResponse>(
      API_ENDPOINTS.patient.todo.delete(todoId),
      {
        method: "DELETE",
        headers: getAuthHeaders(),
      },
      "Failed to delete todo"
    );
  } catch (error) {
    log.error(`Failed to delete todo ${todoId}:`, error);
    throw error;
  }
}

// ============================================================================
// Attachment Services
// ============================================================================

export async function createTimelineAttachment(
  attachmentData: {
    timeline_id: number;
    file_name: string;
    file_type: string;
    file_url: string;
    file_size: number;
  }
): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.attachment.create, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(attachmentData),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to create attachment: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to create attachment:", error);
    throw error;
  }
}

export async function deleteTimelineAttachments(timelineId: number): Promise<ApiSuccessResponse> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.attachment.delete(timelineId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });
 
    if (!response.ok) {
      throw new Error(`Failed to delete timeline attachments: ${response.statusText}`);
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete timeline attachments ${timelineId}:`, error);
    throw error;
  }
}
  

// ============================================================================
// Report Services (New)
// ============================================================================

/**
 * Create a lab report with test items
 */
export async function createLabReport(
  reportData: CreateLabReportRequest
): Promise<{ success: boolean; report_id: number }> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.reports.lab.create, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(reportData),
    });

    if (!response.ok) {
      throw new Error(`Failed to create lab report: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to create lab report:", error);
    throw error;
  }
}

/**
 * Get all lab reports for a timeline
 */
export async function getLabReportsByTimeline(
  timelineId: number
): Promise<LabReport[]> {
  try {
    const response = await fetch(
      API_ENDPOINTS.patient.reports.lab.byTimeline(timelineId),
      {
        method: "GET",
        headers: getAuthHeaders(),
      }
    );

    // If no reports found (404), return empty array instead of throwing error
    if (response.status === 404) {
      return [];
    }

    if (!response.ok) {
      throw new Error(`Failed to get lab reports: ${response.statusText}`);
    }

    const result = await response.json();
    return result.data || [];
  } catch (error) {
    // If error is not a network error and status is 404, return empty array
    if (error instanceof Error && error.message.includes('404')) {
      return [];
    }
    log.error(`Failed to get lab reports for timeline ${timelineId}:`, error);
    throw error;
  }
}

/**
 * Delete a lab report
 */
export async function deleteLabReport(
  reportId: number
): Promise<{ success: boolean }> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.reports.lab.delete(reportId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to delete lab report: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete lab report ${reportId}:`, error);
    throw error;
  }
}

/**
 * Create an imaging report
 */
export async function createImagingReport(
  reportData: CreateImagingReportRequest
): Promise<{ success: boolean; report_id: number }> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.reports.imaging.create, {
      method: "POST",
      headers: {
        ...getAuthHeaders(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(reportData),
    });

    if (!response.ok) {
      throw new Error(`Failed to create imaging report: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error("Failed to create imaging report:", error);
    throw error;
  }
}

/**
 * Get all imaging reports for a timeline
 */
export async function getImagingReportsByTimeline(
  timelineId: number
): Promise<ImagingReport[]> {
  try {
    const response = await fetch(
      API_ENDPOINTS.patient.reports.imaging.byTimeline(timelineId),
      {
        method: "GET",
        headers: getAuthHeaders(),
      }
    );

    // If no reports found (404), return empty array instead of throwing error
    if (response.status === 404) {
      return [];
    }

    if (!response.ok) {
      throw new Error(`Failed to get imaging reports: ${response.statusText}`);
    }

    const result = await response.json();
    return result.data || [];
  } catch (error) {
    // If error is not a network error and status is 404, return empty array
    if (error instanceof Error && error.message.includes('404')) {
      return [];
    }
    log.error(`Failed to get imaging reports for timeline ${timelineId}:`, error);
    throw error;
  }
}

/**
 * Delete an imaging report
 */
export async function deleteImagingReport(
  reportId: number
): Promise<{ success: boolean }> {
  try {
    const response = await fetch(API_ENDPOINTS.patient.reports.imaging.delete(reportId), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to delete imaging report: ${response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    log.error(`Failed to delete imaging report ${reportId}:`, error);
    throw error;
  }
}

// ============================================================================
// Export all services
/**
 * Check if patient with given name already exists (duplicate detection)
 */
async function checkDuplicatePatient(
  name: string,
  exactMatch: boolean = false
): Promise<{ found: boolean; count: number; patients: any[]; search_name: string; exact_match: boolean }> {
  try {
    const url = new URL(API_ENDPOINTS.patient.checkDuplicate, window.location.origin);
    url.searchParams.append("name", name);
    url.searchParams.append("exact_match", String(exactMatch));

    const response = await fetch(url.toString(), {
      method: "GET",
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to check duplicate patient: ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    log.error("Failed to check duplicate patient:", error);
    throw error;
  }
}

// ============================================================================

const patientService = {
  // Patient
  createPatient,
  checkDuplicatePatient,
  getPatient,
  getPatientByEmail,
  listPatients,
  updatePatient,
  deletePatient,

  // Timeline
  createTimelineStage,
  getPatientTimeline,
  getTimelineDetail,
  saveTimelineDetail,
  deleteTimeline,

  // Medical Images
  createMedicalImage,
  deleteTimelineImages,
  // Metrics
  batchCreateMetrics,
  deleteTimelineMetrics,
  // Attachments
  createTimelineAttachment,
  deleteTimelineAttachments,
  // Todos
  createPatientTodo,
  getPatientTodos,
  updateTodoStatus,
  deletePatientTodo,

  // Reports 
  createLabReport,
  getLabReportsByTimeline,
  deleteLabReport,
  createImagingReport,
  getImagingReportsByTimeline,
  deleteImagingReport,
};

export default patientService;

File: backend/services/patient_service.py
import logging
from typing import List, Dict, Optional
from consts.exceptions import AgentRunException
from database import patient_db
 
logger = logging.getLogger(__name__)
 
 
# ============================================================================
# Patient Basic Info Services
# ============================================================================
 
async def create_patient_record(patient_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Create a new patient record
    """
    try:
        # Validate required fields
        required_fields = ['name', 'gender', 'age', 'medical_record_no']
        for field in required_fields:
            if not patient_data.get(field):
                raise ValueError(f"Missing required field: {field}")
 
        result = patient_db.create_patient(patient_data, tenant_id, user_id)
        return {
            "success": True,
            "patient_id": result['patient_id'],
            "message": "Patient created successfully"
        }
    except Exception as e:
        logger.error(f"Failed to create patient: {str(e)}")
        raise AgentRunException(f"Failed to create patient: {str(e)}")
 
 
async def get_patient_info(patient_id: int, tenant_id: str) -> Optional[dict]:
    """
    Get patient information by ID
    """
    try:
        patient = patient_db.get_patient_by_id(patient_id, tenant_id)
        if not patient:
            return None
 
        return patient
    except Exception as e:
        logger.error(f"Failed to get patient info: {str(e)}")
        raise AgentRunException(f"Failed to get patient info: {str(e)}")
 
 
async def get_patient_by_email_service(email: str, tenant_id: str) -> Optional[dict]:
    """
    Get patient information by email address
    """
    try:
        patient = patient_db.get_patient_by_email(email, tenant_id)
        if not patient:
            return None

        return patient
    except Exception as e:
        logger.error(f"Failed to get patient by email: {str(e)}")
        raise AgentRunException(f"Failed to get patient by email: {str(e)}")


async def list_patients_service(tenant_id: str, search_query: Optional[str] = None,
                                filter_type: Optional[str] = None,
                                limit: int = 100, offset: int = 0) -> List[dict]:
    """
    List patients with search and filters
    """
    try:
        patients = patient_db.list_patients(tenant_id, search_query, filter_type, limit, offset)
        return patients
    except Exception as e:
        logger.error(f"Failed to list patients: {str(e)}")
        raise AgentRunException(f"Failed to list patients: {str(e)}")
 
 
async def update_patient_info(patient_id: int, patient_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Update patient information
    """
    try:
        success = patient_db.update_patient(patient_id, patient_data, tenant_id, user_id)
        if not success:
            raise ValueError("Patient not found")
 
        return {
            "success": True,
            "message": "Patient updated successfully"
        }
    except Exception as e:
        logger.error(f"Failed to update patient: {str(e)}")
        raise AgentRunException(f"Failed to update patient: {str(e)}")
 
 
async def delete_patient_record(patient_id: int, tenant_id: str, user_id: str) -> dict:
    """
    Delete patient record (soft delete)
    """
    try:
        success = patient_db.delete_patient(patient_id, tenant_id, user_id)
        if not success:
            raise ValueError("Patient not found")
 
        return {
            "success": True,
            "message": "Patient deleted successfully"
        }
    except Exception as e:
        logger.error(f"Failed to delete patient: {str(e)}")
        raise AgentRunException(f"Failed to delete patient: {str(e)}")
 
 
# ============================================================================
# Patient Timeline Services
# ============================================================================
 
async def create_timeline_stage_service(timeline_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Create a new timeline stage for a patient
    """
    try:
        # Validate required fields
        required_fields = ['patient_id', 'stage_type', 'stage_date', 'stage_title']
        for field in required_fields:
            if not timeline_data.get(field):
                raise ValueError(f"Missing required field: {field}")
 
        result = patient_db.create_timeline_stage(timeline_data, tenant_id, user_id)
        return {
            "success": True,
            "timeline_id": result['timeline_id'],
            "message": "Timeline stage created successfully"
        }
    except Exception as e:
        logger.error(f"Failed to create timeline stage: {str(e)}")
        raise AgentRunException(f"Failed to create timeline stage: {str(e)}")
 
 
async def get_patient_timeline_service(patient_id: int, tenant_id: str) -> List[dict]:
    """
    Get complete timeline for a patient
    """
    try:
        timelines = patient_db.get_patient_timeline(patient_id, tenant_id)
        return timelines
    except Exception as e:
        logger.error(f"Failed to get patient timeline: {str(e)}")
        raise AgentRunException(f"Failed to get patient timeline: {str(e)}")
 
 
async def get_timeline_detail_service(timeline_id: int, tenant_id: str) -> Optional[dict]:
    """
    Get detailed information for a specific timeline stage
    """
    try:
        detail = patient_db.get_timeline_detail(timeline_id, tenant_id)
        return detail
    except Exception as e:
        logger.error(f"Failed to get timeline detail: {str(e)}")
        raise AgentRunException(f"Failed to get timeline detail: {str(e)}")
 
 
async def create_timeline_detail_service(detail_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Create or update timeline detail information
    """
    try:
        if not detail_data.get('timeline_id'):
            raise ValueError("Missing required field: timeline_id")
 
        result = patient_db.create_timeline_detail(detail_data, tenant_id, user_id)
        return {
            "success": True,
            "detail_id": result['detail_id'],
            "message": "Timeline detail saved successfully"
        }
    except Exception as e:
        logger.error(f"Failed to save timeline detail: {str(e)}")
        raise AgentRunException(f"Failed to save timeline detail: {str(e)}")

 
# ============================================================================
# Medical Image Services
# ============================================================================
 
async def upload_medical_image_service(image_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Create a medical image record
    Note: This only creates the database record. Image upload to MinIO should be handled separately.
    """
    try:
        required_fields = ['timeline_id', 'image_type', 'image_label', 'image_url']
        for field in required_fields:
            if not image_data.get(field):
                raise ValueError(f"Missing required field: {field}")
        result = patient_db.create_medical_image(image_data, tenant_id, user_id)
        return {
            "success": True,
            "image_id": result['image_id'],
            "message": "Medical image record created successfully"
        }
    except Exception as e:
        logger.error(f"Failed to create medical image record: {str(e)}")
        raise AgentRunException(f"Failed to create medical image record: {str(e)}")
 
 
# ============================================================================
# Metrics Services
# ============================================================================
 
async def batch_create_metrics_service(metrics_data: List[dict], tenant_id: str, user_id: str) -> dict:
    """
    Batch create patient metrics
    """
    try:
        if not metrics_data or not isinstance(metrics_data, list):
            raise ValueError("metrics_data must be a non-empty list")
 
        result = patient_db.batch_create_metrics(metrics_data, tenant_id, user_id)
        return {
            "success": True,
            "created_count": result['created_count'],
            "message": f"Created {result['created_count']} metrics successfully"
        }
    except Exception as e:
        logger.error(f"Failed to batch create metrics: {str(e)}")
        raise AgentRunException(f"Failed to batch create metrics: {str(e)}")
 
 
# ============================================================================
# Patient Todo Services
# ============================================================================
 
async def create_patient_todo_service(todo_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Create a patient todo item
    """
    try:
        required_fields = ['patient_id', 'todo_title', 'todo_type', 'due_date']
        for field in required_fields:
            if not todo_data.get(field):
                raise ValueError(f"Missing required field: {field}")
 
        result = patient_db.create_patient_todo(todo_data, tenant_id, user_id)
        return {
            "success": True,
            "todo_id": result['todo_id'],
            "message": "Todo created successfully"
        }
    except Exception as e:
        logger.error(f"Failed to create todo: {str(e)}")
        raise AgentRunException(f"Failed to create todo: {str(e)}")
 
 
async def get_patient_todos_service(patient_id: int, tenant_id: str, status: Optional[str] = None) -> List[dict]:
    """
    Get patient todos with optional status filter
    """
    try:
        todos = patient_db.get_patient_todos(patient_id, tenant_id, status)
        return todos
    except Exception as e:
        logger.error(f"Failed to get patient todos: {str(e)}")
        raise AgentRunException(f"Failed to get patient todos: {str(e)}")
 
 
async def update_todo_status_service(todo_id: int, status: str, tenant_id: str, user_id: str) -> dict:
    """
    Update todo status
    """
    try:
        valid_statuses = ['pending', 'completed', 'overdue']
        if status not in valid_statuses:
            raise ValueError(f"Invalid status. Must be one of: {', '.join(valid_statuses)}")

        success = patient_db.update_todo_status(todo_id, status, tenant_id, user_id)

        if not success:
            raise ValueError("Todo not found")

        return {
            "success": True,
            "message": "Todo status updated successfully"
        }
    except Exception as e:
        logger.error(f"Failed to update todo status: {str(e)}")
        raise AgentRunException(f"Failed to update todo status: {str(e)}")

 

 

async def delete_timeline_service(timeline_id: int, tenant_id: str, user_id: str) -> dict:
    """
    Delete a timeline stage (soft delete)
    """
    try:
        success = patient_db.delete_timeline(timeline_id, tenant_id, user_id)

        if not success:
            raise ValueError("Timeline not found")

        return {
            "success": True,
            "message": "Timeline deleted successfully"
        }
    except Exception as e:
        logger.error(f"Failed to delete timeline: {str(e)}")
        raise AgentRunException(f"Failed to delete timeline: {str(e)}")

 

 

async def delete_patient_todo_service(todo_id: int, tenant_id: str, user_id: str) -> dict:
    """
    Delete a patient todo item (soft delete)
    """
    try:
        success = patient_db.delete_patient_todo(todo_id, tenant_id, user_id)

        if not success:
            raise ValueError("Todo not found")

        return {
            "success": True,
            "message": "Todo deleted successfully"
        }
    except Exception as e:
        logger.error(f"Failed to delete todo: {str(e)}")
        raise AgentRunException(f"Failed to delete todo: {str(e)}")


async def delete_timeline_images_service(timeline_id: int, tenant_id: str, user_id: str) -> dict:
    """
    Delete all images for a timeline (soft delete)
    """
    try:
        success = patient_db.delete_timeline_images(timeline_id, tenant_id, user_id)
        return {
            "success": True,
            "message": "Timeline images deleted successfully"
        }
    except Exception as e:
        logger.error(f"Failed to delete timeline images: {str(e)}")
        raise AgentRunException(f"Failed to delete timeline images: {str(e)}")
 
 
async def delete_timeline_metrics_service(timeline_id: int, tenant_id: str, user_id: str) -> dict:
    """
    Delete all metrics for a timeline (soft delete)
    """
    try:
        success = patient_db.delete_timeline_metrics(timeline_id, tenant_id, user_id)
        return {
            "success": True,
            "message": "Timeline metrics deleted successfully"
        }
    except Exception as e:
        logger.error(f"Failed to delete timeline metrics: {str(e)}")
        raise AgentRunException(f"Failed to delete timeline metrics: {str(e)}")


async def create_attachment_service(attachment_data: dict, tenant_id: str, user_id: str) -> dict:
    """
    Create a new attachment record
    """
    try:
        result = patient_db.create_attachment(attachment_data, tenant_id, user_id)
        return {
            "success": True,
            "attachment_id": result['attachment_id'],
            "message": "Attachment created successfully"
        }
    except Exception as e:
        logger.error(f"Failed to create attachment: {str(e)}")
        raise AgentRunException(f"Failed to create attachment: {str(e)}")

async def delete_attachment_service(attachment_id: int, tenant_id: str, user_id: str) -> dict:
    """
    Delete an attachment record
    """
    try:
        success = patient_db.delete_attachment(attachment_id, tenant_id, user_id)
        return {
            "success": True,
            "message": "Attachment deleted successfully"
        }
    except Exception as e:
        logger.error(f"Failed to delete attachment: {str(e)}")
        raise AgentRunException(f"Failed to delete attachment: {str(e)}")